Feature/hpa upscale downscale delay configurable

Dmitry1987 opened the pull request:
**What this PR does / why we need it**:
Makes "upscale forbidden window" and "downscale forbidden window"  duration configurable in arguments of kube-controller-manager. Those are options of horizontal pod autoscaler.

**Special notes for your reviewer**:
Please have a look @DirectXMan12 , the PR as discussed in Slack.

**Release note**:
```
Make "upscale forbidden window" and "downscale forbidden window"  duration configurable in arguments of kube-controller-manager. Those are options of horizontal pod autoscaler. Right now are hardcoded 3 minutes for upscale, and 5 minutes to downscale.  But sometimes cluster administrator might want to change this for his own needs.
```

Dmitry1987 committed:
make hpa upscale and downscale delay window configurable

dannyz29 commented:
+ 1

Dmitry1987 commented:
Hi  @DirectXMan12 ,  one of the commits was made from ubuntu VM where I was testing. And other commits from windows.  So, the commit from ubuntu has an "X", it thinks no CLA signed  (the username of commit "Dmitry Shmakov" different from account username "Dmitry1987"...),  is it ok and will be possible to merge, or I should fix this somehow?

DirectXMan12 commented:
> is it ok and will be possible to merge, or I should fix this somehow?

You'll need to fix it.  You can amend the commit while using the `--author` flag to do so.

DirectXMan12 commented on pkg/apis/componentconfig/types.go:
```diff
@@ -699,6 +699,10 @@ type KubeControllerManagerConfiguration struct {
 	// horizontalPodAutoscalerSyncPeriod is the period for syncing the number of
 	// pods in horizontal pod autoscaler.
 	HorizontalPodAutoscalerSyncPeriod metav1.Duration
+        // horizontalPodAutoscalerUpscaleForbiddenWindow is a period after which next upscale allowed.
```
you appear to have some indentation issues here

DirectXMan12 commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -367,7 +375,7 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 		}
 
-		rescale = shouldScale(hpa, currentReplicas, desiredReplicas, timestamp)
+		rescale = shouldScale(a, hpa, currentReplicas, desiredReplicas, timestamp)
```
just make this a method of the HPA controller instead of passing the HPA as the first argument.

DirectXMan12 reviewed (changes requested):
A couple of minor changes, otherwise looks good in general.  Thanks for doing this! :-)

DirectXMan12 commented:
cc @jszczepkowski

timothysc commented:
Punting to @piosz and @DirectXMan12.  It doesn't appear that sig-instrumentation has their kubernetes team configured properly.

snirad commented:
Very nice !

Dmitry1987 commented:
Did the changes. @DirectXMan12  please have a look :) , 
P.S. thanks for patience and explaining stuff in slack 👍

jszczepkowski commented on pkg/apis/componentconfig/types.go:
```diff
@@ -699,6 +699,10 @@ type KubeControllerManagerConfiguration struct {
+    // horizontalPodAutoscalerUpscaleForbiddenWindow is a period after which next upscale allowed.
+    HorizontalPodAutoscalerUpscaleForbiddenWindow metav1.Duration
+    // horizontalPodAutoscalerDownscaleForbiddenWindow is a period after which next downscale allowed.
+    HorizontalPodAutoscalerDownscaleForbiddenWindow metav1.Duration
```
Wrong indentation: should be tab instead of spaces.

jszczepkowski commented:
Good job!

This issue has been discussed in #32408 (adding the link for reference).

Dmitry1987 commented:
Thanks @jszczepkowski . I've read the thread, but not sure if it means that this PR is needed or not. like, will it be merged, or this functionality will not be accepted?

jszczepkowski commented:
> I've read the thread, but not sure if it means that this PR is needed or not. like, will it be merged, or this functionality will not be accepted?

The PR should be merged (after LGTM) and this functionality will be accepted.

DirectXMan12 commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -155,6 +157,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet, allControllers []string, disabled
 	fs.StringVar(&s.VolumeConfiguration.FlexVolumePluginDir, "flex-volume-plugin-dir", s.VolumeConfiguration.FlexVolumePluginDir, "Full path of the directory in which the flex volume plugin should search for additional third party volume plugins.")
 	fs.Int32Var(&s.TerminatedPodGCThreshold, "terminated-pod-gc-threshold", s.TerminatedPodGCThreshold, "Number of terminated pods that can exist before the terminated pod garbage collector starts deleting terminated pods. If <= 0, the terminated pod garbage collector is disabled.")
 	fs.DurationVar(&s.HorizontalPodAutoscalerSyncPeriod.Duration, "horizontal-pod-autoscaler-sync-period", s.HorizontalPodAutoscalerSyncPeriod.Duration, "The period for syncing the number of pods in horizontal pod autoscaler.")
+	fs.DurationVar(&s.HorizontalPodAutoscalerUpscaleForbiddenWindow.Duration, "horizontal-pod-autoscaler-upscale-delay", s.HorizontalPodAutoscalerUpscaleForbiddenWindow.Duration, "The period since last upscale, in minutes, before another upscale can be performed in horizontal pod autoscaler.")
```
this should probably not have the "in minutes" part.

DirectXMan12 reviewed (changes requested):
Small nit, otherwise LGTM.  Please squash the fixup commits together before merge.

Dmitry1987 reviewed (commented):
commits squashed :)

DirectXMan12 commented:
/lgtm

DirectXMan12 commented:
Whoops, you'll have to update the tests so that they pass
/lgtm cancel

Dmitry1987 commented:
@k8s-bot bazel test this

Dmitry1987 commented:
Can someone please run the tests?  I fixed tests file where it failed last time, and local tests passed, with "./build/run.sh make test"

jszczepkowski commented:
@k8s-bot test this

jszczepkowski commented:
Do we want to merge this on 1.6? Technically, we have a code freeze.

Dmitry1987 commented:
gosh, I updated "horizontal_test.go" but not "legacy_horizontal_test.go"... but my "./build/run.sh make test" passed :(  .. how can I know what are all the test files which may fail?  step by step?

Dmitry1987 commented:
rebased and added needed params to "legacy_horizontal_test.go" (this file didn't yet exist in my branch), so test shouldn't fail now, i hope =) .

jszczepkowski commented:
@k8s-bot test this

Dmitry1987 commented:
Hi guys, from latest failed tests I found and fixed all "gofmt" needed files with "gofmt -s -w".  But why "kops" test failed and where to fix what, I don't understand :( , can someone take a quick look at the log here:  https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/42101/pull-kubernetes-e2e-kops-aws/16450/?log#log
there are errors like ```build/temp.linux-x86_64-2.7/check_libyaml.c:2:18: fatal error: yaml.h: No such file or directory``` which I have no idea how to fix or if they're related to my changes at all (or did those fail as chain reaction following 'bazel build'?).

The 3d test "/hack/verify-flags-underscore.py" failure I fixed too, added flags like it says to "hack/verify-flags/known-flags.txt"...  so I just need help with kops test. Thanks :)

Dmitry1987 commented:
Hi @jszczepkowski , can you help with the bot?  it doesn't run test after my last push... and I don't know what is the failure in 'kops' test, maybe you can help?

jszczepkowski commented:
@k8s-bot test this

DirectXMan12 commented:
as long as the tests pass, looks good!

DirectXMan12 commented:
@k8s-bot ok to test

DirectXMan12 commented:
/lgtm /approve

Dmitry1987 commented:
Hi @deads2k, @mikedanese , do you have a minute to have a look at this PR? 
 It's really small, but we need this feature, and I don't want to rebuild the controller-manager myself with every k8s release we upgrade :P  .  Thanks in advance! :)

deads2k commented:
/approve

mtaufen commented:
/approve
the changes in componentconfig lgtm
